examples/ota-requestor-app/esp32/main/OTARequesterImpl.cpp

```diff
@@ -0,0 +1,282 @@
+        }
+        break;
+    }
+    case 3: { // NotifyUpdateApplied
+        constexpr uint32_t softwareVersion = 1;
+
+        err = cluster.NotifyUpdateApplied(nullptr, nullptr, otaUpdateToken, softwareVersion);
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(SoftwareUpdate, "NotifyUpdateApplied() failed %s", chip::ErrorStr(err));
+        }
+        break;
+    }
+    default:
+        break;
+    }
+}
+
+void OnConnectionFailure(void * context, OperationalDeviceProxy * operationalDeviceProxy, CHIP_ERROR error)
+{
+    ChipLogError(SoftwareUpdate, "failed to connect to: %s", chip::ErrorStr(error));
+}
+
+void OnBlockReceived(void * context, chip::bdx::TransferSession::BlockData & blockdata)
+{
+    if (OTAUpdater::GetInstance().IsInProgress() == false)
+    {
+        OTAUpdater::GetInstance().Begin();
+    }
+    OTAUpdater::GetInstance().Write(reinterpret_cast<const void *>(blockdata.Data), blockdata.Length);
+}
+
+void OnTransferComplete(void * context)
+{
+    ESP_LOGI(TAG, "Transfer complete!");
+    OTAUpdater::GetInstance().End();
+}
+
+void OnTransferFailed(void * context)
+{
+    ESP_LOGI(TAG, "Transfer Failed");
+    OTAUpdater::GetInstance().Abort();
+}
+
+void ConnectToProvider(const char * ipAddress, uint32_t nodeId)
+{
+    // Explicitly calling UpdateAddress() should not be needed once OperationalDeviceProxy can resolve IP address from node ID and
+    NodeId providerNodeId = nodeId;
+    IPAddress ipAddr;
+    IPAddress::FromString(ipAddress, ipAddr);
+    PeerAddress addr = PeerAddress::UDP(ipAddr, CHIP_PORT);
+    gOperationalDeviceProxy.UpdateAddress(addr);
+
+    Server * server                             = &(Server::GetInstance());
+    OperationalDeviceProxyInitParams initParams = {
+        .sessionManager = &(server->GetSecureSessionManager()),
+        .exchangeMgr    = &(server->GetExchangeManager()),
+        .idAllocator    = &(server->GetSessionIDAllocator()),
+        .fabricsTable   = &(server->GetFabricTable()),
+    };
+
+    CHIP_ERROR err              = CHIP_NO_ERROR;
+    FabricIndex peerFabricIndex = providerFabricIndex;
+    gOperationalDeviceProxy.Init(providerNodeId, peerFabricIndex, initParams);
+    err = gOperationalDeviceProxy.Connect(&mOnConnectedCallback, &mOnConnectionFailureCallback);
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(SoftwareUpdate, "Cannot establish connection to peer device: %" CHIP_ERROR_FORMAT, err.Format());
+    }
+}
+
+void OTARequesterImpl::SendQueryImageCommand(const char * ipAddress, uint32_t nodeId)
+{
+    operationalDeviceContext = 1;
+    ConnectToProvider(ipAddress, nodeId);
+}
+
+void OTARequesterImpl::SendApplyUpdateRequestCommand(const char * ipAddress, uint32_t nodeId)
+{
+    operationalDeviceContext = 2;
+    ConnectToProvider(ipAddress, nodeId);
+}
+
+void OTARequesterImpl::SendNotifyUpdateAppliedCommand(const char * ipAddress, uint32_t nodeId)
+{
+    operationalDeviceContext = 2;
```

Comment:
Think this should be 3? Perhaps `operationalDeviceContext` should be an enum so we're not using magic numbers here.